WIP: add a CalendarStorage class and its SqliteCalendarStorage implementation

dcaliste opened the pull request:
This is adding a class linking a KCalendarCore::MemoryCalendar with an SQLiteDB for a given notebook, using a multi-notebook backend.

dcaliste commented:
Rebased on latest #60, but there is also now a test to the new SqliteCalendarStorage class.

pvuorela commented on src/sqlitecalendarstorage.cpp:
```diff
@@ -0,0 +1,255 @@
+
+bool SqliteCalendarStorage::erase()
+{
+    if (notebook() && d->mBackend.deleteNotebook(*notebook())) {
```
I'm not sure about deleting notebooks by "erasing" storage. It feels kind of misleading when the real storage (e.g. a database file)  is not deleted, and commonly there's no use-case to actually need such removal.

dcaliste replied:
As discussed on IRC, I've removed completely the `erase()` routine from the CalendarStorage API.

Since it is used by sync plugins to delete notebooks because of upstream deletion for instance, I've instead created a static Notebook method that delete a "system" notebook.

pvuorela commented on src/sqlitestorage.cpp:
```diff
@@ -368,227 +146,62 @@ bool SqliteStorage::load(const QString &uid)
         }
+        return true;
+    } else {
+        return false;
```
Minor, but could drop the else branch and just return false.

dcaliste replied:
Sure, done.

pvuorela reviewed (commented):
Quite a bit to digest here. Some random comments inline for starters. Overall I feel a bit lost with all these classes and the responsibilities of different Sqlite* classes. Seems there's also some code duplication.

For the usage the buteo-caldav PR is a good reference on the usage. Based on that one thing I'm not sure of is this explicit sqlite references. Earlier the backend was nicely abstracted away with mKCal::ExtendedCalendar::defaultStorage() type of API. Could be nice if it'd be possible to keep something like that.

And speaking of that, this now exports quite many things, though it was somewhat existing setup already. E.g. the current version includes sqliteformat.h in the -devel package which doesn't feel entirely proper. Also some _p.h type of headers included which I doubt the client apps will need.

Did you have a plan for the Extended* classes? Would those go away eventually or remain on the side?

For #60 I suppose these should eventually go in together, or is there any use for sqlite backend class alone? However there are some details which could perhaps be merged separately, e.g. the notebook update time commit is one thing. If enough of those, could merge separately out of the way.

pvuorela commented:
Needs a rebase.

dcaliste commented on src/alarmhandler.h:
```diff
@@ -0,0 +1,60 @@
+  @brief
+  This class provides an interface to handle alarms. 
+*/
+class MKCAL_EXPORT AlarmHandler
```
Yes, it is used in mkcaltool now replacing the ExtendedStorage there.

dcaliste commented on src/calendarhandler_p.h:
```diff
@@ -0,0 +1,85 @@
+
+namespace mKCal {
+
+class MKCAL_HIDE CalendarHandler: public KCalendarCore::Calendar::CalendarObserver
```
Done in a separated commit.

dcaliste commented on src/calendarstorage.cpp:
```diff
@@ -0,0 +1,237 @@
+
+/**
+  Private class that helps to provide binary compatibility between releases.
+  @internal
```
Indeed, dropped.

dcaliste commented on src/calendarstorage.h:
```diff
@@ -0,0 +1,279 @@
+    /**
+      Action to be performed on save for deleted incidences.
+    */
+    enum DeleteAction {
```
It will disappear in ExtendedStorage with it. So Remains in CalendarStorage and MultiCalendarStorage. If I look at the API that is common between the two, I only find this enum. Even the save() routine that is using it has a different API. So I'm not sure it's worth trying to abstract something in common between the two classes.

dcaliste commented on src/multicalendarstorage.h:
```diff
@@ -0,0 +1,331 @@
+/**
+  @file
+  This file is part of the API for handling calendar data and
+  defines the an interface to store a single KCalendarCore::Calendar.
```
Even the text was wrong. Thanks for noticing. Now the description is correct.

dcaliste commented on src/multicalendarstorage.h:
```diff
@@ -0,0 +1,331 @@
+
+/**
+  @brief
+  This class provides a way to agregate different sources of
```
Typo fixed.

pvuorela commented on src/alarmhandler.h:
```diff
@@ -0,0 +1,58 @@
+    virtual KCalendarCore::Incidence::List incidencesWithAlarms(const QString &notebookUid,
+                                                                const QString &uid = QString()) = 0;
+
+ public:
```
Nit: no space before public: or protected:

dcaliste replied:
Corrected here and in calendarhandler_p.h.

pvuorela commented on src/alarmhandler.h:
```diff
@@ -0,0 +1,58 @@
+/**
+  @file
+  This file is part of the API for handling calendar data and
+  defines the an interface to store a single KCalendarCore::Calendar.
```
Some case for "the an" was fixed, but there are still multiple ones copypasting it and the text looks wrong in many of them altogether, like here.

dcaliste replied:
Thanks for noticing. I've cleaned up a bit the documentation strings, and in #60 PR also.

I've added documentation strings for the internal `calendarhandler_p.h` and `alarmhandler_p.h` files also.

pvuorela commented on src/notebook.h:
```diff
@@ -434,6 +434,25 @@ class MKCAL_EXPORT Notebook
+    /**
+      Retrieve the list of notebooks as stored in the system.
+
+      @param account when not null restrict the list to a given
```
Code is using isEmpty(), not isNull(), and string emptiness is a cleaner concept. Could adjust the comment here?

dcaliste replied:
Sure, changed for empty wording.

pvuorela commented on src/multicalendarstorage.cpp:
```diff
@@ -0,0 +1,532 @@
+    const QString notebookUid = ids.first;
+    QString uid = ids.second;
+    if (notebookUid.isEmpty() || uid.isEmpty()) {
+        qCWarning(lcMkcal) << "wrong instance identifier" << instanceIdentifier;
```
Or maybe rather "invalid"? Wrong sounds like unexpected identifier.

dcaliste replied:
Ok, changed for "invalid". It looks better, indeed.

pvuorela replied:
The old text is btw having some other instances later on.

dcaliste replied:
You're right. I've changed the warning label in the three other cases.

pvuorela commented on src/multicalendarstorage.cpp:
```diff
@@ -0,0 +1,532 @@
+    // Even if we're looking for a specific incidence instance, we load all
+    // the series for recurring event, to avoid orphaned exceptions in the
+    // calendar or recurring events without their exceptions.
+    QHash<QString, Incidence::List> hash;
```
Quite minor: could restrict scope by moving into else branch. Up to you.

dcaliste replied:
Ah yes, indeed. Done.

dcaliste commented:
I've also clarified the naming scheme of calendarhandler and alarmhandler files, using the _p for the header and nothing for the cpp.

dcaliste commented:
I've just noticed that the new way to handle alarms is making a large performance regression on database change. I need to rethink how it's done.

dcaliste commented:
I've corrected the performance regression on alarm setting. Maybe it's something that should go in its own PR: the change in alarm handling.

Before, there was the need for the calendar to be present when setting an alarm. Indeed, just the changed incidence is not enough. In case it is recurring, one also needs the recurrence ids of all exceptions for this event. Additionally, if an exception is changed, the parent alarm must also be fetched and reset. It makes things quite complicated in my opinion and requires to have the calendar at hand, which is not necessary something that is easy to pass when we become multi-calendar.

Therefore, I decided to change the way alarm are set and reset, in a simpler way:
- you always set/reset alarms for the full series, even if there is only a change in one component of the series.
- for shake of code simplicity, added, modified and deleted incidences are all treated the same, alarm of the series are removed and then set if the series still exist (the clear step is useless for the added cases and the set case is also useless for the deleted cases, but like that code is the same for all and simpler to read).

The new AlarmHandler class is build on these two ideas. `setupAlarms()` receive a set of (notebookUid, incidenceUid) tuples, clean all alarms for these and then call for the object implementing the AlarmHandler to get the list of full series for these tuples and create alarms accordingly.

In my opinion, the code is much simpler like that, with less risk to get faulty cases because we didn't update properly a parent when an exception is changed or whatever. But it means two "regressions":
- previously, only changed incidences in a series triggered an alarm change. Now, all alarms of the series are cleared out before being reset. Which means for instance that in a recurring event with two exceptions, before when changing one exception, the parent and the exception where cleared and reset. While now, the second exception will also be reset.
- previously, added cases triggered only alarm additions, while now it also makes a DBus call to remove alarms for its uids (which does nothing at the end of course). And these DBus calls may become visible if you add hundreds of events at once.

So I think basically, I can live with these two regressions in my opinion.

Do you think I should make a separated PR for the introduction of the AlarmHandler class ?

pvuorela commented on src/extendedstorage.cpp:
```diff
@@ -434,47 +414,32 @@ bool ExtendedStorage::deleteNotebook(const Notebook::Ptr &nb)
         d->mDefaultNotebook = Notebook::Ptr();
     }
 
+    d->clearAlarms(nb->uid());
```
The intermediate commit was skipping isRunTimeOnly() case. That shouldn't still have alarms? Likewise the setup on alarm handler commit does it for runtime notebooks.

dcaliste replied:
My bad, the if disappeared. I put the test back.

pvuorela commented:
Still scratching my head a bit what's happening with these alarms now or what was going there bad. Seems complicated enough and suppose a bit separate too so possible to have on separate PR without too much hassle? I was first thinking of having everything at once, but everything here is getting somewhat complicated so maybe wouldn't be a bad idea to get some changes again in first. One candidate could be the #60 for some commits first. Dunno.

For "regressions" if that's full series vs incidence and addition becoming from "add alarm" for each to clear and then add alarms for each, suppose that's only doubling the calls still

If you're interested in avoiding time spend on timed dbus, some of those sync calls could be replaced easily with async as the result is only checked for reporting failures. There's even an ancient task for using async api, JB#14882.

dcaliste commented on src/alarmhandler.cpp:
```diff
@@ -0,0 +1,296 @@
+
+    QMap<QString, QVariant> query;
+    query["APPLICATION"] = "libextendedkcal";
+    if (uids.count() == 1) {
```
I've separated the routine into the two cases:
- one that deletes from a given notebook, where one can save a DBus call,
- one that deletes from random incidences.

The first case is used when changing the visibility of a notebook, or deleting it. The second is used when modifying incidences (adding, updating or removing).

I've used a different list that cookies for the second case. For clarity.

dcaliste commented:
I've rebased the PR on #65 and #60. And I've added the listing routines for sync in the new CalendarStorage class.
